### PR TITLE
EOS-28559: Add and delete new access key

### PR DESCRIPTION
**Pull request by abhijit1patil**
# Problem Statement
- Add and delete new access key

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design
[Wiki Add access key](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/900530404/Add+Access+Key+REST+API+Specifications)

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/core/blogic/const.py [hunk 1]
```diff
@@ -930,6 +930,7 @@
 RGW_ADMIN_OPERATIONS_MAPPING_SCHEMA = '{}/schema/rgw_admin_api_operations.json'.format(CSM_PATH)
 CREATE_USER_OPERATION = 'CREATE_USER'
 RGW_CLIENT_ERROR_MSG = 'Can not process the request'
+ADD_ACCESS_KEY_OPERATION = 'ADD_ACCESS_KEY'
```
Comment (rohitkolapkar) [inline csm/core/blogic/const.py:933]: Keep the operation names similar to rgw ops. 
ADD_ACCESS_KEY_OPERATION -> CREATE_KEY

File: csm/core/blogic/const.py [hunk 2]
```diff
@@ -943,3 +944,4 @@
 RGW_JSON_MAX_BUCKETS = 'max-buckets'
 RGW_JSON_SUSPENDED = 'suspended'
 RGW_JSON_TENANT = 'tenant'
+RGW_JSON_SUBUSER = 'subuser'
```
Comment (rohitkolapkar) [inline csm/core/blogic/const.py:947]: Minor: 'subuser' is not supported, skip this request body param.

File: csm/core/services/rgw/s3/users.py [hunk 3]
```diff
@@ -43,3 +43,17 @@ async def create_user(self, **user_body):
         if isinstance(plugin_response, RgwError):
             self._handle_error(plugin_response)
         return plugin_response
+
+    @Log.trace_method(Log.DEBUG, exclude_args=['access_key', 'secret_key'])
+    async def add_access_key(self, **add_access_key_body):
```
Comment (rohitkolapkar) [inline csm/core/services/rgw/s3/users.py:48]: please rename: add_access_key -> create_key

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

File: csm/core/controllers/rgw/s3/users.py [hunk 4]
```diff
@@ -79,3 +79,56 @@ async def put(self):
...
+            add_access_key_body = schema.load(await self.request.json(), unknown='EXCLUDE')
+            Log.debug(f"Handling create s3 iam user PUT request"
+                  f" request body: {add_access_key_body}")
+        except json.decoder.JSONDecodeError:
+            raise InvalidRequest(message_args="Invalid Request Body")
+        except ValidationError as val_err:
+            raise InvalidRequest(f"{ValidationErrorFormatter.format(val_err)}")
+        with self._guard_service():
+            response = await self._service.add_access_key(**add_access_key_body)
+            return CsmResponse(response)
```
[the remark this answers is not in the document]
Comment (abhijit1patil) [reply csm/core/controllers/rgw/s3/users.py]: updated

File: csm/core/services/rgw/s3/users.py (hunk 3 above)
Comment (abhijit1patil) [reply csm/core/services/rgw/s3/users.py]: renamed to create_key

File: csm/core/controllers/rgw/s3/users.py [hunk 5]
```diff
@@ -79,3 +79,56 @@ async def put(self):
         with self._guard_service():
             response = await self._service.create_user(**user_body)
             return CsmResponse(response)
+
+class AddAccessKeySchema(Schema):
```
[the remark this answers is not in the document]
Comment (abhijit1patil) [reply csm/core/controllers/rgw/s3/users.py]: updated

File: csm/core/controllers/rgw/s3/users.py [hunk 6]
```diff
@@ -79,3 +111,58 @@ async def put(self):
...
+
+    PUT: Add/Create access key
+    """
+
+    def __init__(self, request):
+        """S3 IAM User List View Init."""
+        super().__init__(request)
+        self._service = self.request.app[const.RGW_S3_IAM_USERS_SERVICE]
+
+    @CsmAuth.permissions({Resource.S3_IAM_USERS: {Action.UPDATE}})
```
Comment (rohitkolapkar) [inline csm/core/controllers/rgw/s3/users.py:128]: What should be the appropriate Action ? update or create ? 
if update, then u might need to give update access rights to 's3_iam_users' resource.

File: csm/core/controllers/rgw/s3/users.py [hunk 7]
```diff
@@ -79,3 +79,56 @@ async def put(self):
...
+    @Log.trace_method(Log.DEBUG)
+    async def put(self):
+        """
+        PUT REST implementation for creating a new s3 iam user.
+        """
```
[the remark this answers is not in the document]
Comment (abhijit1patil) [reply csm/core/controllers/rgw/s3/users.py]: updated

**Pull request closed without merging**